### PR TITLE
feat: support adding multiple users in a single message

### LINKED ISSUE #100: Support adding multiple users in a single message (Use NLP parsing per sentence, not on an entire message)
**Is your feature request related to a problem? Please describe.**
It would be great to have a way to bulk add users in a single request.

**Describe the solution you'd like**
Implementing per sentence parsing instead of per-message will allow us to do this. It will also help the NLP to be more effective in determining intentions. 

**Describe alternatives you've considered**

**Additional context**
Case of a user adding lots of users: https://github.com/zeroDevs/coding_challenge-13/pull/161

Broken case on message too big:
https://github.com/zeroDevs/coding_challenge-13/issues/132#issuecomment-457773736
screen shot 2019-01-25 at 6 48 53 pm

Case of not working:
See https://github.com/freecodecamp-ibadan/freecodecamp-ibadan/issues/4
> `@all-contributors please add @stevoo24 for code,content`
screen shot 2019-02-10 at 11 30 04 am

### PR DESCRIPTION
**Pull request by jakebolam**
Closes #100, #146 and https://github.com/all-contributors/all-contributors/issues/591

- additional improve NLP parsing of usernames

### CODE REVIEW

File: src/tasks/processIssueComment/probot-processIssueComment.js [hunk 1]
```diff
@@ -133,14 +162,18 @@ async function probotProcessIssueComment({ context, commentReply, analytics }) {
...
+        analytics.track('addContributors', {
             who: commentBody,
-            contributions,
+            contributors,
         })
-        const safeWho = getSafeRef(who)
+
+        const whos = contributors.map(({ who }) => who).join('--')
+        const safeWho = getSafeRef(whos)
+        // TODO: max length on branch name?
```
Comment (Berkmann18) [inline src/tasks/processIssueComment/probot-processIssueComment.js]: If we were to have a max length, what would it be?

File: src/tasks/processIssueComment/probot-processIssueComment.js [hunk 2]
```diff
@@ -15,35 +15,55 @@ const {
...
+        return addContributor((contributor.who, contributor.contributions))
     })
 
+    await Promise.all(addContributorPromises)
+
+    if (usersAdded === 0) {
+        return commentReply.reply(
+        `I couldn't determine any contributions for ${usersAdded.join(', ')}.
+        Did you specify any contributions? Please make sure to use [valid contribution names](https://allcontributors.org/docs/en/emoji-key).`,
+    )
```
Comment (Berkmann18) [inline src/tasks/processIssueComment/probot-processIssueComment.js]: Is this block supposed to finish here or are the following lines in that `if` block no longer useful?

**Comment by Berkmann18** [comment]
Do you think this should be supported by the CLI too?
Cf. https://github.com/all-contributors/all-contributors-cli/issues/187
That way the bot could have an easier time adding multiple users.

**Comment by zeke** [comment]
👋 Hey I think this would be a cool feature to land, and I would love to use it on some of my team's repos that already have a lot of existing contributors. Can I help move this PR along somehow?

**Comment by gr2m** [comment]
Hey there, I plan to look into it once I find some time to familiarize myself with the code base, I've blocked a day for that on Sep 16th, sorry it takes so long :(

**Comment by Berkmann18** [comment]
@gr2m Did you managed to get something working on this?

**Comment by gr2m** [comment]
I biggest problem is that the code base is very complex, it is using Probot in a serverless environment which the current version is not build for. With the next version of Probot that should be much simpler, so I focus on that first, because then it would be allow to spend some of my work hours to migrate `all-contributors` to it as a means of testing.

The current code base is too hard to maintain and contribute to.

**Comment by Berkmann18** [comment]
Oh okay, the codebase is very complex. I wonder how we could go about simplifying it.

File: src/tasks/processIssueComment/probot-processIssueComment.js (hunk 1 above)
Comment (jsoref) [reply src/tasks/processIssueComment/probot-processIssueComment.js:175]: I'd suggest 255 chars. While it's true that people could limit themselves to using GitHub's ui for everything, and with recent changes, it's possible to [renaming existing branches](https://github.com/github/renaming#renaming-existing-branches), branch names are generally stored in file systems and can be subject to the whims of file systems, some of whom don't particularly like very long file names (I'm not suggesting an 11 character limit...).

I suspect that would give you ~16-20 contributors, which is quite a lot.

Comment (Berkmann18) [reply src/tasks/processIssueComment/probot-processIssueComment.js:175]: @jsoref That makes sense, yeah 255 chars seems good to me.

**Comment by tenshiAMD** [comment]
@jakebolam @Berkmann18 @gr2m I think we can revisit this one.

**Comment by Berkmann18** [comment]
> @jakebolam @Berkmann18 @gr2m I think we can revisit this one.

Yup, how would you go about it

**Review by Berkmann18 — COMMENTED** [review]
That looks good to me; although there are commented-out tests like in `parse-comment/index.test.js`, my earlier review comments still hold and it might be refreshing tests to fix that coverage error.

**Comment by tenshiAMD** [comment]
> That looks good to me (although there are commented-out tests like in `parse-comment/index.test.js`).

@Berkmann18 those are still WIP

**Comment by tenshiAMD** [comment]
Related to https://github.com/all-contributors/app/issues/210

**Comment by tenshiAMD** [comment]
@Berkmann18 @gr2m almost done 🚀

File: test/tasks/processIssueComment/utils/parse-comment/index.test.js [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    //         ),
+    //     ).toEqual({
+    //         action: 'add',
+    //         contributors: {
+    //             kazydek: ['doc', 'review', 'maintenance'],
+    //             akucharska: ['code', 'maintenance'],
+    //             derberg: ['doc', 'ideas']
+    //         },
+    //     })
+    // })
```
Comment (tenshiAMD) [inline test/tasks/processIssueComment/utils/parse-comment/index.test.js:228]: @gr2m @Berkmann18 do we need to include this one? I think this is an invalid test for me. Let me know your thoughts about this. Thanks.

Comment (gr2m) [reply test/tasks/processIssueComment/utils/parse-comment/index.test.js:228]: I'd like to wait for @Berkmann18 to comment, he knows more about the natural language processing

Comment (tenshiAMD) [reply test/tasks/processIssueComment/utils/parse-comment/index.test.js:228]: > I'd like to wait for @Berkmann18 to comment, he knows more about the natural language processing

@gr2m ok sure 😄 there are only 2 tests left then we can merge this one 🚀

Comment (tenshiAMD) [reply test/tasks/processIssueComment/utils/parse-comment/index.test.js:228]: @Berkmann18 please check and let me know. Also, if you want to add other tests we need for this one. Cheers! 🎉

**Comment by tenshiAMD** [comment]
Fixes https://github.com/all-contributors/all-contributors/issues/375.

File: test/tasks/processIssueComment/utils/parse-comment/index.test.js (hunk 3 above)
Comment (tenshiAMD) [reply test/tasks/processIssueComment/utils/parse-comment/index.test.js:228]: @Berkmann18 gentle ping

Comment (gr2m) [reply test/tasks/processIssueComment/utils/parse-comment/index.test.js:228]: okay let's go ahead and merge, we can revert if Max finds any problems

**Review by gr2m — APPROVED** [review]

File: test/tasks/processIssueComment/utils/parse-comment/index.test.js (hunk 3 above)
Comment (tenshiAMD) [reply test/tasks/processIssueComment/utils/parse-comment/index.test.js:228]: > okay let's go ahead and merge, we can revert if Max finds any problems

@gr2m all right, let's merge this one 🚀

**Review by tenshiAMD — APPROVED** [review]

**Pull request merged**